src/app/qgstemplateprojectsmodel.cpp

```diff
@@ -0,0 +1,103 @@
+ *                                                                         *
+ ***************************************************************************/
+
+#include "qgstemplateprojectsmodel.h"
+#include "qgsziputils.h"
+#include "qgssettings.h"
+#include "qgsapplication.h"
+#include "qgis.h"
+#include "qgsrecentprojectsitemsmodel.h"
+
+#include <QStandardPaths>
+#include <QDir>
+#include <QCryptographicHash>
+#include <QPainter>
+
+#include <memory>
+
+
+QgsTemplateProjectsModel::QgsTemplateProjectsModel( QObject *parent )
+  : QStandardItemModel( parent )
+{
+  const QStringList paths = QStandardPaths::standardLocations( QStandardPaths::AppDataLocation );
+  QString templateDirName = QgsSettings().value( QStringLiteral( "qgis/projectTemplateDir" ),
+                            QgsApplication::qgisSettingsDirPath() + QStringLiteral( "project_templates" ) ).toString();
+
+  for ( const QString &templatePath : paths )
+  {
+    const QString path = templatePath + QDir::separator() + QStringLiteral( "project_templates" );
+    scanDirectory( path );
+    mFileSystemWatcher.addPath( path );
+  }
+
+  scanDirectory( templateDirName );
+  mFileSystemWatcher.addPath( templateDirName );
+
+  connect( &mFileSystemWatcher, &QFileSystemWatcher::directoryChanged, this, &QgsTemplateProjectsModel::scanDirectory );
+
+  setColumnCount( 1 );
+}
+
+void QgsTemplateProjectsModel::scanDirectory( const QString &path )
+{
+  QDir dir = QDir( path );
+  const QFileInfoList files = dir.entryInfoList( QStringList() << QStringLiteral( "*.qgs" ) << QStringLiteral( "*.qgz" ) );
+
+  // Remove any template from this directory)
+  for ( int i = rowCount() - 1; i >= 0; --i )
+  {
+    if ( index( i, 0 ).data( QgsRecentProjectItemsModel::NativePathRole ).toString().startsWith( path ) )
+    {
+      removeRow( i );
+    }
+  }
+
+  // Refill with templates from this directory
+  for ( const QFileInfo &file : files )
+  {
+    std::unique_ptr<QStandardItem> item = qgis::make_unique<QStandardItem>( file.fileName() ) ;
+
+    const QString fileId = QCryptographicHash::hash( file.filePath().toUtf8(), QCryptographicHash::Sha224 ).toHex();
+
+    QStringList files;
+    QDir().mkpath( mTemporaryDir.filePath( fileId ) );
+
+    QgsZipUtils::unzip( file.filePath(), mTemporaryDir.filePath( fileId ), files );
+
+    QString filename( mTemporaryDir.filePath( fileId ) + QDir::separator() + QStringLiteral( "preview.png" ) );
+
+    QImage thumbnail( filename );
+    if ( !thumbnail.isNull() )
+    {
+      //nicely round corners so users don't get paper cuts
+      QImage previewImage( thumbnail.size(), QImage::Format_ARGB32 );
+      previewImage.fill( Qt::transparent );
+      QPainter previewPainter( &previewImage );
+      previewPainter.setRenderHint( QPainter::Antialiasing, true );
+      previewPainter.setPen( Qt::NoPen );
+      previewPainter.setBrush( Qt::black );
+      previewPainter.drawRoundedRect( 0, 0, previewImage.width(), previewImage.height(), 8, 8 );
```

Comment:
Hi dpi compatibility?

Reply:
Compatible with the recent projects and with the image size. Unless something is changed https://github.com/qgis/QGIS/pull/10020#discussion_r284921565 here and aligned in recent projects, I wouldn't touch that.